contracts/hooks/FullRange.sol

```diff
@@ -0,0 +1,469 @@
+        // transfer liquidity tokens to erc20 contract
+        UniswapV4ERC20 erc20 = UniswapV4ERC20(poolToERC20[key.toId()]);
+
+        erc20.transferFrom(msg.sender, address(0), liquidity);
+
+        IPoolManager.ModifyPositionParams memory params = IPoolManager.ModifyPositionParams({
+            tickLower: MIN_TICK,
+            tickUpper: MAX_TICK,
+            liquidityDelta: -int256(liquidity)
+        });
+
+        modifyPosition(key, params);
+
+        // here, all of the necessary liquidity should have been removed, this portion is just to update fees and feeGrowth
+        HookPosition storage position = poolToHookPosition[key.toId()];
+
+        uint128 positionLiquidity = position.liquidity;
+        require(positionLiquidity >= liquidity);
+
+        Position.Info memory posInfo = poolManager.getPosition(key.toId(), address(this), MIN_TICK, MAX_TICK);
+        // TODO: ?
+        // uint256 feeGrowthInside0LastX128 = posInfo.feeGrowthInside0LastX128;
+        // uint256 feeGrowthInside1LastX128 = posInfo.feeGrowthInside1LastX128;
+
+        position.tokensOwed0 += uint128(
+            FullMath.mulDiv(
+                posInfo.feeGrowthInside0LastX128 - position.feeGrowthInside0LastX128,
+                positionLiquidity,
+                FixedPoint128.Q128
+            )
+        );
+        position.tokensOwed1 += uint128(
+            FullMath.mulDiv(
+                posInfo.feeGrowthInside1LastX128 - position.feeGrowthInside1LastX128,
+                positionLiquidity,
+                FixedPoint128.Q128
+            )
+        );
+
+        position.feeGrowthInside0LastX128 = posInfo.feeGrowthInside0LastX128;
+        position.feeGrowthInside1LastX128 = posInfo.feeGrowthInside1LastX128;
+        // subtraction is safe because we checked positionLiquidity is gte liquidity
+        position.liquidity = uint128(positionLiquidity - liquidity);
+    }
+
+    // deploy ERC-20 contract
+    function beforeInitialize(address, IPoolManager.PoolKey calldata key, uint160) external override returns (bytes4) {
+        require(key.tickSpacing == 60, "Tick spacing must be default");
+
+        // deploy erc20 contract
+
+        bytes memory bytecode = type(UniswapV4ERC20).creationCode;
+        bytes32 salt = keccak256(abi.encodePacked(key.toId()));
+
+        address poolToken;
+        assembly {
+            poolToken := create2(0, add(bytecode, 32), mload(bytecode), salt)
+        }
+
+        poolToERC20[key.toId()] = poolToken;
+
+        return FullRange.beforeInitialize.selector;
+    }
+
+    function _rebalance(IPoolManager.PoolKey calldata key) internal {
+        console.log("we're rebalancing");
+        // retrieve all liquidity -- i think i should already have the liquidity kept track of
+        // uint128 hookLiquidity = poolManager.getLiquidity(key.toId(), address(this), MIN_TICK, MAX_TICK);
+
+        // position.liquidity = hookLiquidity;
+
+        HookPosition storage position = poolToHookPosition[key.toId()];
```

Comment:
I think you'll save gas by making this `memory`, then just writing it to storage when it's ready to be stored.